R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+          paste(package$version)]]$package
+    }
+    ## Now combine the dependencies from this package with our big dependency
+    ## list.
+
+    current_dependencies <- combine_dependencies(
+      single_package_dependencies
+      , current_dependencies
+      , package$name)
+  }
+
+  if (identical(master_list, current_dependencies)) return(master_list)
+
+  ## Run through the entire list again, because we have a bunch of new packages
+  get_dependencies_for_list(current_dependencies, lock, previously_parsed_deps)
+}
+
+#' Attach the latest available lockbox version to a package
+add_latest_version_in_lockbox <- function(package) {
+  package$latest_version_in_lockbox <- max_package_version(
+    list.files(file.path(lockbox_library(), package)))
+  package
+}
+
+#' Take a vector of versions and find the max without coercing to package_version
+max_package_version <- function(versions) {
+  if (length(versions) == 0) return(NULL)
+  formatted_versions <- package_version(as.character(versions))
+  versions[which(formatted_versions == max(formatted_versions))[1]]
+}
+
+#' Check a dependency list for inclusion in the lockfile and replace the package
+#' with the locked version if it does appear there.  Also throw
+#' an error if we require a dependency version greater than that specified by
+#' the lockfile.
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- Filter(function(l) l$name == package$name, lock)[[1]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
+          , " is required by package ", package$parent_package
+          , ", but lockbox is locked at version: "
+          , as.character(locked_package$version)
+          , ". Please update your lockfile accordingly"))
+    }
+    package <- locked_package
+  } else {
+    package$is_dependency_package <- TRUE
+  }
+  if (!"remote" %in% names(package) || is.na(package$remote)) {
+    package$remote <- "CRAN"
+  }
+  package <- as.locked_package(package)
+  if (package$is_dependency_package && !package$name %in%
+    as.character(installed.packages(priority = "base")[,1])) {
+      package$latest_version <- package$latest_version_in_lockbox %||%
+        package$latest_version %||% get_latest_version(package)
+  }
+  package
+}
+
+get_latest_version <- function(package) {
+  if (package$remote == "CRAN") {
+    get_available_cran_version(package)
+  } else {
+    version_from_remote(package)
+  }
+}
+
+#' Combine two lists of dependencies via version comparisons.  Keep packages
+#' found in list1 on the left side of the entirety of list2, while moving
+#' the parent package to the space after it's rightmost dependency found in list2.
```

Comment:
Ok I think I understand now. So you're saying that after combining, the final order of installation will be left-to-right, so you are "bubbling" the packages with dependencies to their right so that they occur later in the final sequence? I'm thinking about whether there is a cleaner, more understandable way. One way that might not be feasible given R's lack of big integer support is:

1. Tag each dependency-less package with an increasing integer sequence.
2. Tag each 1-dependency package with the sequence, continually increasing.
etc.

Until we've run out of packages. Then we just `order()` by the tags! Is that doable?